### ISSUE TITLE
Restore password?

### ISSUE THREAD

**Issue by neutronscott** [freeotp/freeotp-android#344]
I don't recall setting or being asked for a password for FreeOTP. I created a backup, and now want to restore. What's the password?

I think it would be best to confirm or set the password during the backup operation.

**Comment by jpopovitch** [comment]
When you first install FreeOTP it asks for a password and instructs the user to retain and remember the password because it will be needed to restore any future backup.

**Comment by WebSnke** [comment]
> When you first install FreeOTP it asks for a password and instructs the user to retain and remember the password because it will be needed to restore any future backup.

@neutronscott 
Yep, there is no way that FreeOTP didn't ask you for a password. You should try to restore your 2FA accounts through the backup codes that you were supposed to write down when you enabled 2FA.

**Comment by neutronscott** [comment]
I've been using it for years. It was restored. This is merely a request to make the UX better.

**Comment by WebSnke** [comment]
> I've been using it for years. It was restored. This is merely a request to make the UX better.

Ok, now I get it. You suggest having to enter your password a second time when setting it up, right?

**Comment by neutronscott** [comment]
Yes. A reminder. One may blindly export and with a false sense of confidence factory reset their phone only to release their fault when restoring.

**Comment by Profpatsch** [comment]
Wow, I’ve just been bit by this. This is horrible UX! These keys are some of the most important pieces of data, being able to export them without hidden traps is absolutely vital.

Unless this is fixed soon, I won’t be recommending this app anymore and switch to a different OTP store.

**Comment by neutronscott** [comment]
I think that's sufficient. Bonus if can be reset.

**Comment by VincentSC** [comment]
I really don't like these "blame the user" answers here.  :(

**Comment by neutronscott** [comment]
Well, maybe it's time to try writing java. I'm sure the team would welcome a PR.

**Comment by VincentSC** [comment]
If there is a roadmap for the general design I believe in, then there are several recruiters that believe my Java skills are till there. I already looked and I think I could even do it. :)

In case there is no roadmap, should I make a PR for that first?

**Comment by joggee-fr** [comment]
@VincentSC, it seems PRs are very welcome here. Feel free to add one for this topic.

**Comment by emptyhead** [comment]
Just to add to this. If there's not an important technical reason to have the password set up during installation, setting up at the point of backup would be a much better experience. Password is then fresh in user's mind and it also means they'll understand that a password will be needed to restore the backup.

**Comment by ShadyMedic** [comment]
I agree with emptyhead, I was in a rush when I was setting the app up for the first time. I used some bogus password that I didn't bother remembering for more then a few minutes, knowing that I have backup codes for everything and hopefully I'll be able to set a proper password for any later backup exported from the app. Now, it seems like I'll have to reset 2FA in all the services I set set it up, one by one because I can't move the keys from my current installation.

I simply don't understand why does the password for future (and non-automatic) backups needs to be set long before any backup is created and cannot be changed at a later time.

The requirement to set the password before it's needed leads to users reusing a password they have elsewhere. If they could set a password right before exporting a backup that will likely be imported into another installation within a few minutes, the users would be more likely to use a unique password that they'll remember for those few minutes and forgeting it afterwards, making the backup file useless to 3rd parties in case of it being leaked.

**Comment by VincentSC** [comment]
> (...) knowing that I have backup codes for everything (...)

That triggered me to remember some old lessons about UX...

It's all about wording in the UX:
- a **password** one can lose
- a **backup code** one needs to restore

When designing a frontend, and you assume that the user will assume the same assumption, then we get into these problems.

So I suggest stopping to use the word "password" ASAP, to prevent others from finding this tread.
